File: articles/mathematical-modeling-in-public-health-and-epidemiology/cpp/epidemic_scenario_summary.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

struct EpidemicScenario {
    std::string key;
    std::string scenario_name;
    double population;
    double initial_infectious;
    double initial_recovered;
    double beta;
    double gamma;
    int days;
    double hospital_capacity;
    double hospitalization_rate;
};

struct Evaluation {
    double r0_simple;
    double peak_infectious;
    double peak_hospital_demand;
    double capacity_margin;
    bool capacity_breach;
};

Evaluation evaluate(const EpidemicScenario& scenario) {
    double susceptible = scenario.population - scenario.initial_infectious - scenario.initial_recovered;
    double infectious = scenario.initial_infectious;
    double recovered = scenario.initial_recovered;
    double peak_infectious = infectious;
    double peak_hospital_demand = infectious * scenario.hospitalization_rate;

    for (int day = 0; day < scenario.days; ++day) {
        const double new_infections = scenario.beta * susceptible * infectious / scenario.population;
        const double new_recoveries = scenario.gamma * infectious;
        susceptible = std::max(0.0, susceptible - new_infections);
        infectious = std::max(0.0, infectious + new_infections - new_recoveries);
        recovered = std::min(scenario.population, recovered + new_recoveries);
        peak_infectious = std::max(peak_infectious, infectious);
        peak_hospital_demand = std::max(peak_hospital_demand, infectious * scenario.hospitalization_rate);
    }

    return {
        scenario.beta / scenario.gamma,
        peak_infectious,
        peak_hospital_demand,
        scenario.hospital_capacity - peak_hospital_demand,
        peak_hospital_demand > scenario.hospital_capacity
    };
}

int main() {
    std::vector<EpidemicScenario> scenarios = {
        {"baseline", "Baseline transmission", 100000.0, 120.0, 4000.0, 0.32, 0.12, 120, 850.0, 0.045},
        {"moderate_intervention", "Moderate intervention", 100000.0, 120.0, 4000.0, 0.24, 0.12, 120, 850.0, 0.045},
        {"strong_intervention", "Strong intervention", 100000.0, 120.0, 4000.0, 0.18, 0.12, 120, 850.0, 0.045},
        {"vaccination_plus_intervention", "Vaccination plus intervention", 100000.0, 120.0, 22000.0, 0.20, 0.12, 120, 850.0, 0.030}
    };

    std::cout << "key,r0_simple,peak_infectious,peak_hospital_demand,capacity_margin,capacity_breach\n";
    for (const auto& scenario : scenarios) {
        const Evaluation eval = evaluate(scenario);
        std::cout << std::fixed << std::setprecision(6)
                  << scenario.key << "," << eval.r0_simple << ","
                  << eval.peak_infectious << "," << eval.peak_hospital_demand << ","
                  << eval.capacity_margin << ","
                  << (eval.capacity_breach ? "true" : "false") << "\n";
    }

    return 0;
}

File: articles/mathematical-modeling-in-public-health-and-epidemiology/fortran/epidemic_scenario_summary.f90
program epidemic_scenario_summary
  implicit none

  integer, parameter :: n = 4
  character(len=40) :: keys(n)
  real(8) :: population(n), initial_infectious(n), initial_recovered(n)
  real(8) :: beta(n), gamma(n), hospital_capacity(n), hospitalization_rate(n)
  real(8) :: r0(n), peak_infectious(n), peak_hospital(n), capacity_margin(n)
  integer :: days(n), i
  logical :: breach

  keys = (/ 'baseline                                ', 'moderate_intervention                   ', &
            'strong_intervention                     ', 'vaccination_plus_intervention           ' /)
  population = (/ 100000.0d0, 100000.0d0, 100000.0d0, 100000.0d0 /)
  initial_infectious = (/ 120.0d0, 120.0d0, 120.0d0, 120.0d0 /)
  initial_recovered = (/ 4000.0d0, 4000.0d0, 4000.0d0, 22000.0d0 /)
  beta = (/ 0.32d0, 0.24d0, 0.18d0, 0.20d0 /)
  gamma = (/ 0.12d0, 0.12d0, 0.12d0, 0.12d0 /)
  days = (/ 120, 120, 120, 120 /)
  hospital_capacity = (/ 850.0d0, 850.0d0, 850.0d0, 850.0d0 /)
  hospitalization_rate = (/ 0.045d0, 0.045d0, 0.045d0, 0.030d0 /)

  print '(A)', 'key r0_simple peak_infectious peak_hospital_demand capacity_margin capacity_breach'

  do i = 1, n
    call evaluate_sir(population(i), initial_infectious(i), initial_recovered(i), beta(i), gamma(i), &
      days(i), hospital_capacity(i), hospitalization_rate(i), r0(i), peak_infectious(i), peak_hospital(i), capacity_margin(i))
    breach = peak_hospital(i) > hospital_capacity(i)
    print '(A,1X,F8.4,1X,F12.3,1X,F12.3,1X,F12.3,1X,L1)', trim(keys(i)), &
      r0(i), peak_infectious(i), peak_hospital(i), capacity_margin(i), breach
  end do

contains

  subroutine evaluate_sir(population, initial_infectious, initial_recovered, beta, gamma, days, &
      hospital_capacity, hospitalization_rate, r0_simple, peak_infectious, peak_hospital_demand, capacity_margin)
    implicit none
    real(8), intent(in) :: population, initial_infectious, initial_recovered, beta, gamma
    integer, intent(in) :: days
    real(8), intent(in) :: hospital_capacity, hospitalization_rate
    real(8), intent(out) :: r0_simple, peak_infectious, peak_hospital_demand, capacity_margin
    real(8) :: susceptible, infectious, recovered, new_infections, new_recoveries, hospital_demand
    integer :: day

    susceptible = population - initial_infectious - initial_recovered
    infectious = initial_infectious
    recovered = initial_recovered
    peak_infectious = infectious
    peak_hospital_demand = infectious * hospitalization_rate

    do day = 1, days
      new_infections = beta * susceptible * infectious / population
      new_recoveries = gamma * infectious
      susceptible = max(0.0d0, susceptible - new_infections)
      infectious = max(0.0d0, infectious + new_infections - new_recoveries)
      recovered = min(population, recovered + new_recoveries)
      peak_infectious = max(peak_infectious, infectious)
      hospital_demand = infectious * hospitalization_rate
      peak_hospital_demand = max(peak_hospital_demand, hospital_demand)
    end do

    r0_simple = beta / gamma
    capacity_margin = hospital_capacity - peak_hospital_demand
  end subroutine evaluate_sir

end program epidemic_scenario_summary
